Tensor type set

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25308 Tensor type set**

Instead of storing a single TensorTypeId in a Tensor, we store a bitset of tensor type IDs in a Tensor, TensorTypeSet. This class comes with some unit tests.  This is in preparation for making Variable a TensorTypeId. In order to help flush out places where this makes a semantic difference, we rename `Tensor::type_id()` to `Tensor::type_set()` and smoke out all of the locations where this was semantically meaningful.

Because the new tensor type set is 64-bits, this increases the size of Tensor by a word.

Listing of semantic changes:
* Many TensorImpl related constructors just propagate TensorTypeId to a parent constructor. These are pretty simple to adjust.
  * Backend extensions are now in the business of explicitly constructing a TensorTypeSet and then passing it in. This is probably OK for now but when Variable drops, these dispatch IDs may get immediately overwritten to have Variable set.
* `sparseTensorSetToDeviceType` and similar functions previously did an equality test with TensorTypeId, to determine what an appropriate device type is. This equality is now replaced with a set inclusion test. This is valid, under the assumption that we don't ever have weird sets like "this tensor is simultaneously a sparse CPU tensor and a sparse CUDA tensor", which will be true in the short term plan of adding Variable to the dispatch ID.
* `impl::dispatchTypeId` was generally introduced for cases where we legitimately need to convert from `TensorTypeSet -> TensorTypeId` in a dispatch related manner. At the moment, the implementation is trivial, but they will soon be adjusted to handle TLS. I've tried to make these call sites as forwards compatible as possible:
  * `checked_tensor_unwrap` and co now use `dispatchTypeId`. When Variable is added to the type set, these will always be called in a context where the Variable type ID is disabled, so we will get the correct underlying tensor type ID.
  * Uses of `Backend` in dispatch are now replaced with `TensorTypeSet`. The general heuristic here for whether or not to accept a `TensorTypeId` or `TensorTypeSet` is that we want to make the generated code as simple as possible. It is easier to retrieve a `TensorTypeSet`, so that's a more appropriate API in these cases.
* In some cases, I could not conveniently switch an implementation to the new semantics, because it was blocked on some other refactor. In this case, I introduced `legacyExtractTypeId`, which gives what would be a BC-compatible `TensorTypeSet` to `TensorTypeId` implementation that will continue to report the same values it would have prior to this change. This is **different** from `dispatchTypeId`, because this function does NOT respect TLS; it always ignores Variable type IDs.
  * c10 dispatcher tests, which are oblivious to Variable dispatch, use this BC function (actually, they use `extractTypeId`, an overload for Tensor.
  * The implementation of `new_*` methods heavily relies on tensor type ID, I chose not to unwind this. PR to refactor this at https://github.com/pytorch/pytorch/pull/25475
  * Slicing also relies on tensor type ID, see `torch/csrc/autograd/python_variable_indexing.cpp` (though in some cases in this file, I was able to replace use of tensor type ID with TensorOptions)
* In some cases, there is an equality test on tensor type ID which would be better done by testing "tensor axes". In those cases, I replaced those equality tests with more equality tests.
  * Example: `torch/csrc/nn/type_checks.h`
  * There is a total punt in `torch/csrc/tensor/python_tensor.cpp` where "instance of" checking is done via dispatch ids. In general, the Variable-ness of a tensor doesn't participate in instanceof testing. It's not entirely clear what to do here.
  * Instead of storing `Backend` in `VariableInfo`, we now just store Layout.

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang committed:
Tensor type set

Instead of storing a single TensorTypeId in a Tensor, we store a bitset
of tensor type IDs in a Tensor.  This is in preparation for making
Variable a TensorTypeId

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791/)

ezyang committed:
Update on "Tensor type set"


Instead of storing a single TensorTypeId in a Tensor, we store a bitset
of tensor type IDs in a Tensor.  This is in preparation for making
Variable a TensorTypeId

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791/)

ezyang committed:
Update on "Tensor type set"


Instead of storing a single TensorTypeId in a Tensor, we store a bitset
of tensor type IDs in a Tensor.  This is in preparation for making
Variable a TensorTypeId

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang committed:
Update on "Tensor type set"


Instead of storing a single TensorTypeId in a Tensor, we store a bitset
of tensor type IDs in a Tensor.  This is in preparation for making
Variable a TensorTypeId

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang committed:
leading zeros, not trailing zeros, LOL on "Tensor type set"


Instead of storing a single TensorTypeId in a Tensor, we store a bitset
of tensor type IDs in a Tensor.  This is in preparation for making
Variable a TensorTypeId

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang commented:
This is ready for review

ezyang commented:
Clang tidy warning is spurious

ezyang commented:
Somehow, allreduce failure is real:

```
Sep 03 15:57:52 test_allreduce_coalesced_checks (__main__.ProcessGroupGlooTest) ... ERROR:root:Caught exception: Torch not compiled with CUDA enabled, exiting process with exit code: 10
Sep 03 15:57:52 ERROR:root:Caught exception: Torch not compiled with CUDA enabled, exiting process with exit code: 10
Sep 03 15:57:52 ERROR:root:Caught exception: Torch not compiled with CUDA enabled, exiting process with exit code: 10
Sep 03 15:57:52 ERROR:root:Caught exception: Torch not compiled with CUDA enabled, exiting process with exit code: 10
Sep 03 15:57:52 FAIL
Sep 03 15:57:52 test_allreduce_coalesced_stress (__main__.ProcessGroupGlooTest) ... ERROR:root:Caught exception: le() received an invalid combination of arguments - got (unicode), but expected one of:
Sep 03 15:57:52  * (Tensor other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52  * (Number other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52 , exiting process with exit code: 10
Sep 03 15:57:52 ERROR:root:Caught exception: le() received an invalid combination of arguments - got (unicode), but expected one of:
Sep 03 15:57:52  * (Tensor other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52  * (Number other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52 , exiting process with exit code: 10
Sep 03 15:57:52 ERROR:root:Caught exception: le() received an invalid combination of arguments - got (unicode), but expected one of:
Sep 03 15:57:52  * (Tensor other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52  * (Number other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52 , exiting process with exit code: 10
Sep 03 15:57:52 ERROR:root:Caught exception: le() received an invalid combination of arguments - got (unicode), but expected one of:
Sep 03 15:57:52  * (Tensor other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52  * (Number other)
Sep 03 15:57:52       didn't match because some of the arguments have invalid types: (!unicode!)
Sep 03 15:57:52 , exiting process with exit code: 10
Sep 03 15:57:53 FAIL
```

ezyang commented:
Oh, maybe I just need a rebase

ezyang committed:
Update on "Tensor type set"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25308 Tensor type set**

Instead of storing a single TensorTypeId in a Tensor, we store a bitset of tensor type IDs in a Tensor, TensorTypeSet. This class comes with some unit tests.  This is in preparation for making Variable a TensorTypeId. In order to help flush out places where this makes a semantic difference, we rename `Tensor::type_id()` to `Tensor::type_set()` and smoke out all of the locations where this was semantically meaningful.

Because the new tensor type set is 64-bits, this increases the size of Tensor by a word.

Listing of semantic changes:
* Many TensorImpl related constructors just propagate TensorTypeId to a parent constructor. These are pretty simple to adjust.
  * Backend extensions are now in the business of explicitly constructing a TensorTypeSet and then passing it in. This is probably OK for now but when Variable drops, these dispatch IDs may get immediately overwritten to have Variable set.
* `sparseTensorSetToDeviceType` and similar functions previously did an equality test with TensorTypeId, to determine what an appropriate device type is. This equality is now replaced with a set inclusion test. This is valid, under the assumption that we don't ever have weird sets like "this tensor is simultaneously a sparse CPU tensor and a sparse CUDA tensor", which will be true in the short term plan of adding Variable to the dispatch ID.
* `impl::dispatchTypeId` was generally introduced for cases where we legitimately need to convert from `TensorTypeSet -> TensorTypeId` in a dispatch related manner. At the moment, the implementation is trivial, but they will soon be adjusted to handle TLS. I've tried to make these call sites as forwards compatible as possible:
  * `checked_tensor_unwrap` and co now use `dispatchTypeId`. When Variable is added to the type set, these will always be called in a context where the Variable type ID is disabled, so we will get the correct underlying tensor type ID.
  * Uses of `Backend` in dispatch are now replaced with `TensorTypeSet`. The general heuristic here for whether or not to accept a `TensorTypeId` or `TensorTypeSet` is that we want to make the generated code as simple as possible. It is easier to retrieve a `TensorTypeSet`, so that's a more appropriate API in these cases.
* In some cases, I could not conveniently switch an implementation to the new semantics, because it was blocked on some other refactor. In this case, I introduced `legacyExtractTypeId`, which gives what would be a BC-compatible `TensorTypeSet` to `TensorTypeId` implementation that will continue to report the same values it would have prior to this change. This is **different** from `dispatchTypeId`, because this function does NOT respect TLS; it always ignores Variable type IDs.
  * c10 dispatcher tests, which are oblivious to Variable dispatch, use this BC function (actually, they use `extractTypeId`, an overload for Tensor.
  * The implementation of `new_*` methods heavily relies on tensor type ID, I chose not to unwind this. PR to refactor this at https://github.com/pytorch/pytorch/pull/25475
  * Slicing also relies on tensor type ID, see `torch/csrc/autograd/python_variable_indexing.cpp` (though in some cases in this file, I was able to replace use of tensor type ID with TensorOptions)
* In some cases, there is an equality test on tensor type ID which would be better done by testing "tensor axes". In those cases, I replaced those equality tests with more equality tests.
  * Example: `torch/csrc/nn/type_checks.h`
  * There is a total punt in `torch/csrc/tensor/python_tensor.cpp` where "instance of" checking is done via dispatch ids. In general, the Variable-ness of a tensor doesn't participate in instanceof testing. It's not entirely clear what to do here.
  * Instead of storing `Backend` in `VariableInfo`, we now just store Layout.

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang commented:
Based on experience in #25597 I think it was a mistake to make the TensorImpl constructor accept a TensorTypeSet rather than a TensorTypeId. This is because you should not inherit Variable-ness by default: Variable-ness is triggered solely by whether or not `autograd_meta_` is null or not. Most of the time this doesn't matter because we subsequently propagate autograd meta, but there are a few places where it does (`test_embedding_bag`). I'll do a closer audit on this soon.

zdevito reviewed (commented):
The TypeSet class looks good. Changes look reasonable. I do not want to unilaterally approve because I think someone closer to the TensorImpl/TensorTypeId logic needs to check more about the details. I like the stub methods which distinguish intent (i.e. get the legacyTypeID vs get the dispatchTypeID) eventhough both currently do the same thing.

ezyang committed:
rebase on "Tensor type set"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25308 Tensor type set**

Instead of storing a single TensorTypeId in a Tensor, we store a bitset of tensor type IDs in a Tensor, TensorTypeSet. This class comes with some unit tests.  This is in preparation for making Variable a TensorTypeId. In order to help flush out places where this makes a semantic difference, we rename `Tensor::type_id()` to `Tensor::type_set()` and smoke out all of the locations where this was semantically meaningful.

Because the new tensor type set is 64-bits, this increases the size of Tensor by a word.

Listing of semantic changes:
* Many TensorImpl related constructors just propagate TensorTypeId to a parent constructor. These are pretty simple to adjust.
  * Backend extensions are now in the business of explicitly constructing a TensorTypeSet and then passing it in. This is probably OK for now but when Variable drops, these dispatch IDs may get immediately overwritten to have Variable set.
* `sparseTensorSetToDeviceType` and similar functions previously did an equality test with TensorTypeId, to determine what an appropriate device type is. This equality is now replaced with a set inclusion test. This is valid, under the assumption that we don't ever have weird sets like "this tensor is simultaneously a sparse CPU tensor and a sparse CUDA tensor", which will be true in the short term plan of adding Variable to the dispatch ID.
* `impl::dispatchTypeId` was generally introduced for cases where we legitimately need to convert from `TensorTypeSet -> TensorTypeId` in a dispatch related manner. At the moment, the implementation is trivial, but they will soon be adjusted to handle TLS. I've tried to make these call sites as forwards compatible as possible:
  * `checked_tensor_unwrap` and co now use `dispatchTypeId`. When Variable is added to the type set, these will always be called in a context where the Variable type ID is disabled, so we will get the correct underlying tensor type ID.
  * Uses of `Backend` in dispatch are now replaced with `TensorTypeSet`. The general heuristic here for whether or not to accept a `TensorTypeId` or `TensorTypeSet` is that we want to make the generated code as simple as possible. It is easier to retrieve a `TensorTypeSet`, so that's a more appropriate API in these cases.
* In some cases, I could not conveniently switch an implementation to the new semantics, because it was blocked on some other refactor. In this case, I introduced `legacyExtractTypeId`, which gives what would be a BC-compatible `TensorTypeSet` to `TensorTypeId` implementation that will continue to report the same values it would have prior to this change. This is **different** from `dispatchTypeId`, because this function does NOT respect TLS; it always ignores Variable type IDs.
  * c10 dispatcher tests, which are oblivious to Variable dispatch, use this BC function (actually, they use `extractTypeId`, an overload for Tensor.
  * The implementation of `new_*` methods heavily relies on tensor type ID, I chose not to unwind this. PR to refactor this at https://github.com/pytorch/pytorch/pull/25475
  * Slicing also relies on tensor type ID, see `torch/csrc/autograd/python_variable_indexing.cpp` (though in some cases in this file, I was able to replace use of tensor type ID with TensorOptions)
* In some cases, there is an equality test on tensor type ID which would be better done by testing "tensor axes". In those cases, I replaced those equality tests with more equality tests.
  * Example: `torch/csrc/nn/type_checks.h`
  * There is a total punt in `torch/csrc/tensor/python_tensor.cpp` where "instance of" checking is done via dispatch ids. In general, the Variable-ness of a tensor doesn't participate in instanceof testing. It's not entirely clear what to do here.
  * Instead of storing `Backend` in `VariableInfo`, we now just store Layout.

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

gchanan commented on c10/core/TensorTypeSet.h:
```diff
@@ -0,0 +1,84 @@
+    return static_cast<bool>(repr_ & TensorTypeSet(t).repr_);
+  }
+  // Perform set union
+  TensorTypeSet operator|(TensorTypeSet other) const {
```
is there precedence for this kind of design?  I thought given you have functions called add, remove, etc. this would be called union?  Or, everything would be Bit based and you'd just expose those operators.

ezyang replied:
Python sets overload bitwise operators to mean union, so I'm taking a page from their API book. A semantic model of TensorTypeSet as a bit set is pretty accurate, and I think these overloads help.

gchanan replied:
ya, I think the only weirdness to me is that you also don't provide `union` in addition to `operator|` (this is minor).  Python always gives you names, and also overloads the operators in places they make sense.  It feels like a mixed metaphor if sometimes we have names, and sometimes we have overloaded operators.

ezyang replied:
Going to fix your comment, I realized another reason I didn't add a union method: union is a reserved keyword in C++, so I can't :)

gchanan replied:
that seems like a good reason :).

gchanan commented on c10/core/TensorTypeSet.h:
```diff
@@ -0,0 +1,84 @@
+    return TensorTypeSet(repr_ & ~TensorTypeSet(t).repr_);
+  }
+  // Is the set empty?  (AKA undefined tensor)
+  bool empty() const {
```
I'm not sure I totally understand where we still use undefined tensors, but it feels a little strange to provide this as a BitSet interface when there is this outlier with semantic meaning.

ezyang replied:
This is not explained clearly in the patch (yet!), but here is how I think of it.

What we call `TensorTypeId` is actually `optional<RealTensorTypeId>`. The `nullopt` case corresponds to `UndefinedTensorId`. `firstTypeId` returns an `optional<RealTensorTypeId>` because there may not be any RealTensorTypeIds in the set: so you have to return a `nullopt` (aka `UndefinedTensorId` in that case). So, the mapping is:

* `UndefinedTensorId ==> empty TensorTypeSet`
* `CPUTensorId ==> singleton TensorTypeSet containing only CPUTensorId`

`UndefinedTensor` never actually is present in the bitset.

gchanan replied:
you don't have to do this, but it seems nice to do an inventory of places where we actually use UndefinedTensorId.  I bet we can rid of most/all of them by using optional and then we wouldn't have to do this.

gchanan commented on c10/core/TensorTypeSet.h:
```diff
@@ -0,0 +1,84 @@
+  }
+  // Return the "first" type id in this set; i.e., the one that
+  // should handle dispatch for this operator
+  TensorTypeId firstTypeId() const {
```
this is kind of scary from the API perspective (maybe you get rid of this later?).  But clearly the order of TensorTypeId matters, but that's totally opaque here.  Why is this different from dispatchTypeId?

ezyang replied:
> maybe you get rid of this later?

No, it always exists as piece of API on TensorTypeSet.

> But clearly the order of TensorTypeId matters, but that's totally opaque here.

I can add this to the comment.

> Why is this different from dispatchTypeId?

firstTypeId is a pure function on the bitset: it's always well-defined what the first bit in the bitset is. However, dispatchTypeId must account for TLS when determining what the actual dispatch should be. So that lives external to this class. (Underlying design principle: if you write a data structure, put only pure functions on the data structure, and let external code handle interaction with the external world, e.g., TLS)

ezyang replied:
Renaming this to highestPriorityTypeId

gchanan commented on c10/core/TensorTypeSet.h:
```diff
@@ -0,0 +1,84 @@
+// variable? + profiling? + tracing? + lazy? + tensor type id.
+//
+// An undefined tensor is one with an empty tensor type set.
+class TensorTypeSet {
```
bikeshedding: if this is a set of TensorTypeIds why is this not TensorTypeIdSet?

Disclaimer: I think "Type" should be banned in the same way "Backend" should be banned: people use the term to mean 10 different things, so we shouldn't name anything _actually_ that.

ezyang replied:
Mostly because TensorTypeIdSet looks more ugly to me :)

Based on more recent design docs, we kind of want something like `TensorTypeId => DispatchKey` and then `TensorTypeSet => DispatchKeySet`. But that would have involved a lot of annoying renaming, which I wanted to avoid for the PRs.

gchanan replied:
the end state sounds good to me though.

ezyang commented on c10/core/TensorTypeSet.h:
```diff
@@ -0,0 +1,84 @@
+// For the forseeable future, it will still be possible to extract /that/
+// TensorTypeId, and that's what this function does.
+//
+// TODO: this will need to change when Variable drops.
```
Mixtape drop :)

ezyang commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -115,8 +115,9 @@ Tensor new_with_storage(c10::TensorTypeId type_id, at::ScalarType scalar_type, S
 Tensor new_with_tensor(c10::TensorTypeId type_id, at::ScalarType scalar_type, const Tensor& other) {
-  if (other.type_id() != type_id) {
-    throw TypeError("expected %s (got %s)", type_id, other.type_id());
+  if (!other.type_set().has(type_id)) {
```
`legacyExtract` would have been suitable here. In fact, since this is legacy code, I should probably use `legacyExtract` here.

The reasoning for why TensorTypeId equality test should turn into set equality is something like this:

1. We want to be semantics preserving
2. Eventually, some of these type sets are going to contain variable. But the old test would have still passed even if the tensor was a variable. So id equality to set equality is wrong
3. What's a set operation that does work in this case? Set membership.

When I wrote the patch, I didn't have `legacyExtract`; eventually I added it because some of the cases were too annoying to reason out.

ezyang replied:
The semantic difference between these two, is that `set.has(VariableTensorId)` can succeed, but `legacyExtract(set) == VariableTensorId` is always false.

ezyang committed:
Update on "Tensor type set"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25308 Tensor type set**

Instead of storing a single TensorTypeId in a Tensor, we store a bitset of tensor type IDs in a Tensor, TensorTypeSet. This class comes with some unit tests.  This is in preparation for making Variable a TensorTypeId. In order to help flush out places where this makes a semantic difference, we rename `Tensor::type_id()` to `Tensor::type_set()` and smoke out all of the locations where this was semantically meaningful.

Because the new tensor type set is 64-bits, this increases the size of Tensor by a word.

Listing of semantic changes:
* Many TensorImpl related constructors just propagate TensorTypeId to a parent constructor. These are pretty simple to adjust.
  * Backend extensions are now in the business of explicitly constructing a TensorTypeSet and then passing it in. This is probably OK for now but when Variable drops, these dispatch IDs may get immediately overwritten to have Variable set.
* `sparseTensorSetToDeviceType` and similar functions previously did an equality test with TensorTypeId, to determine what an appropriate device type is. This equality is now replaced with a set inclusion test. This is valid, under the assumption that we don't ever have weird sets like "this tensor is simultaneously a sparse CPU tensor and a sparse CUDA tensor", which will be true in the short term plan of adding Variable to the dispatch ID.
* `impl::dispatchTypeId` was generally introduced for cases where we legitimately need to convert from `TensorTypeSet -> TensorTypeId` in a dispatch related manner. At the moment, the implementation is trivial, but they will soon be adjusted to handle TLS. I've tried to make these call sites as forwards compatible as possible:
  * `checked_tensor_unwrap` and co now use `dispatchTypeId`. When Variable is added to the type set, these will always be called in a context where the Variable type ID is disabled, so we will get the correct underlying tensor type ID.
  * Uses of `Backend` in dispatch are now replaced with `TensorTypeSet`. The general heuristic here for whether or not to accept a `TensorTypeId` or `TensorTypeSet` is that we want to make the generated code as simple as possible. It is easier to retrieve a `TensorTypeSet`, so that's a more appropriate API in these cases.
* In some cases, I could not conveniently switch an implementation to the new semantics, because it was blocked on some other refactor. In this case, I introduced `legacyExtractTypeId`, which gives what would be a BC-compatible `TensorTypeSet` to `TensorTypeId` implementation that will continue to report the same values it would have prior to this change. This is **different** from `dispatchTypeId`, because this function does NOT respect TLS; it always ignores Variable type IDs.
  * c10 dispatcher tests, which are oblivious to Variable dispatch, use this BC function (actually, they use `extractTypeId`, an overload for Tensor.
  * The implementation of `new_*` methods heavily relies on tensor type ID, I chose not to unwind this. PR to refactor this at https://github.com/pytorch/pytorch/pull/25475
  * Slicing also relies on tensor type ID, see `torch/csrc/autograd/python_variable_indexing.cpp` (though in some cases in this file, I was able to replace use of tensor type ID with TensorOptions)
* In some cases, there is an equality test on tensor type ID which would be better done by testing "tensor axes". In those cases, I replaced those equality tests with more equality tests.
  * Example: `torch/csrc/nn/type_checks.h`
  * There is a total punt in `torch/csrc/tensor/python_tensor.cpp` where "instance of" checking is done via dispatch ids. In general, the Variable-ness of a tensor doesn't participate in instanceof testing. It's not entirely clear what to do here.
  * Instead of storing `Backend` in `VariableInfo`, we now just store Layout.

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang commented:
Review comments addressed! I also noticed that there are some small hunks in the next patch that might be better on this patch; I can move them around if people want me to.

bwasti commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -15,6 +16,26 @@
+// TODO: I'm not sure if this should live in this header or not; the operant
+// question is whether or not we have access to all the relevant TLS at this
+// point.
+static inline TensorTypeId dispatchTypeId(TensorTypeSet ts) {
```
why is this a separate function hidden away here? wouldn't it either be a method in Tensor or just in ts.highestPriorityTypeId?

smessmer replied:
Separate header probably makes more sense because you can include it from globalATenDispatch **and** the c10 dispatcher.

ezyang replied:
In the later PRs, I determined that yes, I do in fact have access to all the TLS :)

Otherwise, there's no particular reason I put it here, except that this was where I needed the first site. The function *should* be hidden away, since only the dispatchers really have any legitimate use for it.

ezyang replied:
If it turns out I need to move it for c10 dispatcher, I'll move it, but for now I intend to park it here.

bwasti reviewed (approved):
This looks good to me
(commented on previous version of PR in diff and you responded)

ezyang committed:
Update on "Tensor type set"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25308 Tensor type set**

Instead of storing a single TensorTypeId in a Tensor, we store a bitset of tensor type IDs in a Tensor, TensorTypeSet. This class comes with some unit tests.  This is in preparation for making Variable a TensorTypeId. In order to help flush out places where this makes a semantic difference, we rename `Tensor::type_id()` to `Tensor::type_set()` and smoke out all of the locations where this was semantically meaningful.

Because the new tensor type set is 64-bits, this increases the size of Tensor by a word.

Listing of semantic changes:
* Many TensorImpl related constructors just propagate TensorTypeId to a parent constructor. These are pretty simple to adjust.
  * Backend extensions are now in the business of explicitly constructing a TensorTypeSet and then passing it in. This is probably OK for now but when Variable drops, these dispatch IDs may get immediately overwritten to have Variable set.
* `sparseTensorSetToDeviceType` and similar functions previously did an equality test with TensorTypeId, to determine what an appropriate device type is. This equality is now replaced with a set inclusion test. This is valid, under the assumption that we don't ever have weird sets like "this tensor is simultaneously a sparse CPU tensor and a sparse CUDA tensor", which will be true in the short term plan of adding Variable to the dispatch ID.
* `impl::dispatchTypeId` was generally introduced for cases where we legitimately need to convert from `TensorTypeSet -> TensorTypeId` in a dispatch related manner. At the moment, the implementation is trivial, but they will soon be adjusted to handle TLS. I've tried to make these call sites as forwards compatible as possible:
  * `checked_tensor_unwrap` and co now use `dispatchTypeId`. When Variable is added to the type set, these will always be called in a context where the Variable type ID is disabled, so we will get the correct underlying tensor type ID.
  * Uses of `Backend` in dispatch are now replaced with `TensorTypeSet`. The general heuristic here for whether or not to accept a `TensorTypeId` or `TensorTypeSet` is that we want to make the generated code as simple as possible. It is easier to retrieve a `TensorTypeSet`, so that's a more appropriate API in these cases.
* In some cases, I could not conveniently switch an implementation to the new semantics, because it was blocked on some other refactor. In this case, I introduced `legacyExtractTypeId`, which gives what would be a BC-compatible `TensorTypeSet` to `TensorTypeId` implementation that will continue to report the same values it would have prior to this change. This is **different** from `dispatchTypeId`, because this function does NOT respect TLS; it always ignores Variable type IDs.
  * c10 dispatcher tests, which are oblivious to Variable dispatch, use this BC function (actually, they use `extractTypeId`, an overload for Tensor.
  * The implementation of `new_*` methods heavily relies on tensor type ID, I chose not to unwind this. PR to refactor this at https://github.com/pytorch/pytorch/pull/25475
  * Slicing also relies on tensor type ID, see `torch/csrc/autograd/python_variable_indexing.cpp` (though in some cases in this file, I was able to replace use of tensor type ID with TensorOptions)
* In some cases, there is an equality test on tensor type ID which would be better done by testing "tensor axes". In those cases, I replaced those equality tests with more equality tests.
  * Example: `torch/csrc/nn/type_checks.h`
  * There is a total punt in `torch/csrc/tensor/python_tensor.cpp` where "instance of" checking is done via dispatch ids. In general, the Variable-ness of a tensor doesn't participate in instanceof testing. It's not entirely clear what to do here.
  * Instead of storing `Backend` in `VariableInfo`, we now just store Layout.

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

smessmer commented on c10/core/TensorTypeSet.h:
```diff
@@ -0,0 +1,107 @@
+// handling code if one of the inputs requires grad.)
+//
+// An undefined tensor is one with an empty tensor type set.
+class TensorTypeSet {
```
nit: can be final

smessmer commented on c10/core/TensorTypeSet.h:
```diff
@@ -0,0 +1,107 @@
+    return static_cast<bool>(repr_ & TensorTypeSet(t).repr_);
+  }
+  // Perform set union
+  TensorTypeSet operator|(TensorTypeSet other) const {
```
I think `operator|` is the wrong thing here, it exposes the implementation detail of it being a bitfield in the API. Semantically, these are sets. Let's call it `merge()` or `add()` or `append()`?

ezyang replied:
I disagree. Python defines set union to be overloaded with `|`, so this usage is consistent.

smessmer reviewed (approved):
Generally looks good. Do you have benchmarks for these changes? I don't feel comfortable about landing these without.

ezyang commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -82,7 +82,10 @@ static PyObject* Tensor_instancecheck(PyTensorType* self, PyObject* arg) {
+    //
+    // TODO: Actually, this might be bad.  Because Variable-ness
+    // XXXX THIS IS BAD
+    if (var.type_set().highestPriorityTypeId() == self->get_type_id() &&
```
@gchanan says we should deprecate this

ezyang committed:
Update on "Tensor type set"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25308 Tensor type set**

Instead of storing a single TensorTypeId in a Tensor, we store a bitset of tensor type IDs in a Tensor, TensorTypeSet. This class comes with some unit tests.  This is in preparation for making Variable a TensorTypeId. In order to help flush out places where this makes a semantic difference, we rename `Tensor::type_id()` to `Tensor::type_set()` and smoke out all of the locations where this was semantically meaningful.

Because the new tensor type set is 64-bits, this increases the size of Tensor by a word.

Listing of semantic changes:
* Many TensorImpl related constructors just propagate TensorTypeId to a parent constructor. These are pretty simple to adjust.
  * Backend extensions are now in the business of explicitly constructing a TensorTypeSet and then passing it in. This is probably OK for now but when Variable drops, these dispatch IDs may get immediately overwritten to have Variable set.
* `sparseTensorSetToDeviceType` and similar functions previously did an equality test with TensorTypeId, to determine what an appropriate device type is. This equality is now replaced with a set inclusion test. This is valid, under the assumption that we don't ever have weird sets like "this tensor is simultaneously a sparse CPU tensor and a sparse CUDA tensor", which will be true in the short term plan of adding Variable to the dispatch ID.
* `impl::dispatchTypeId` was generally introduced for cases where we legitimately need to convert from `TensorTypeSet -> TensorTypeId` in a dispatch related manner. At the moment, the implementation is trivial, but they will soon be adjusted to handle TLS. I've tried to make these call sites as forwards compatible as possible:
  * `checked_tensor_unwrap` and co now use `dispatchTypeId`. When Variable is added to the type set, these will always be called in a context where the Variable type ID is disabled, so we will get the correct underlying tensor type ID.
  * Uses of `Backend` in dispatch are now replaced with `TensorTypeSet`. The general heuristic here for whether or not to accept a `TensorTypeId` or `TensorTypeSet` is that we want to make the generated code as simple as possible. It is easier to retrieve a `TensorTypeSet`, so that's a more appropriate API in these cases.
* In some cases, I could not conveniently switch an implementation to the new semantics, because it was blocked on some other refactor. In this case, I introduced `legacyExtractTypeId`, which gives what would be a BC-compatible `TensorTypeSet` to `TensorTypeId` implementation that will continue to report the same values it would have prior to this change. This is **different** from `dispatchTypeId`, because this function does NOT respect TLS; it always ignores Variable type IDs.
  * c10 dispatcher tests, which are oblivious to Variable dispatch, use this BC function (actually, they use `extractTypeId`, an overload for Tensor.
  * The implementation of `new_*` methods heavily relies on tensor type ID, I chose not to unwind this. PR to refactor this at https://github.com/pytorch/pytorch/pull/25475
  * Slicing also relies on tensor type ID, see `torch/csrc/autograd/python_variable_indexing.cpp` (though in some cases in this file, I was able to replace use of tensor type ID with TensorOptions)
* In some cases, there is an equality test on tensor type ID which would be better done by testing "tensor axes". In those cases, I replaced those equality tests with more equality tests.
  * Example: `torch/csrc/nn/type_checks.h`
  * There is a total punt in `torch/csrc/tensor/python_tensor.cpp` where "instance of" checking is done via dispatch ids. In general, the Variable-ness of a tensor doesn't participate in instanceof testing. It's not entirely clear what to do here.
  * Instead of storing `Backend` in `VariableInfo`, we now just store Layout.

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang commented:
@smessmer Comprehensive benchmarks for the full bench are at https://github.com/pytorch/pytorch/pull/25653#issue-314106967

gchanan commented on aten/src/ATen/core/op_registration/test_helpers.h:
```diff
@@ -79,3 +79,9 @@ void expectListEquals(c10::ArrayRef<T> expected, std::vector<T> actual) {
 }
+
+// NB: This is not really sound, but all of the type sets constructed here
+// are singletons so it's fine
```
should we assert this?

ezyang replied:
Since this is test code I'm not too concerned about it.

gchanan replied:
ya, I meant `legacyExtractTypeId` more generally -- is that sound?

ezyang replied:
Yeah, `legacyExtractTypeId` is sound. The unsoundness that I'm remarking about here, is that arguably we SHOULD be testing against the entirety of the type set, not just a single type id. But since there's no variables in the c10 tests, it doesn't matter.

gchanan reviewed (approved):
implementation looks good.

One thing that would be nice is a writeup somewhere about the plan, e.g. I never saw it explained in this PR whether we expect the type set to change.

Also, (as discussed in person) I don't think it makes sense to take a TypeSet as a constructor arg to TensorImpl, we should just take a TensorTypeId, as before.  Speaking of which, perhaps we should have some verification code somewhere that your set makes sense?  Like, you don't pass SparseTensorCPU and SparseTensorCUDA?

ezyang commented:
> One thing that would be nice is a writeup somewhere about the plan, e.g. I never saw it explained in this PR whether we expect the type set to change.

The answer is, yes! (And it does, in the next PR)

> Also, (as discussed in person) I don't think it makes sense to take a TypeSet as a constructor arg to TensorImpl, we should just take a TensorTypeId, as before. Speaking of which, perhaps we should have some verification code somewhere that your set makes sense? Like, you don't pass SparseTensorCPU and SparseTensorCUDA?

The answer also is yes! We should make it take a single argument.

ezyang committed:
Update on "Tensor type set"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25308 Tensor type set**

Instead of storing a single TensorTypeId in a Tensor, we store a bitset of tensor type IDs in a Tensor, TensorTypeSet. This class comes with some unit tests.  This is in preparation for making Variable a TensorTypeId. In order to help flush out places where this makes a semantic difference, we rename `Tensor::type_id()` to `Tensor::type_set()` and smoke out all of the locations where this was semantically meaningful.

Because the new tensor type set is 64-bits, this increases the size of Tensor by a word.

Listing of semantic changes:
* Many TensorImpl related constructors just propagate TensorTypeId to a parent constructor. These are pretty simple to adjust.
  * Backend extensions are now in the business of explicitly constructing a TensorTypeSet and then passing it in. This is probably OK for now but when Variable drops, these dispatch IDs may get immediately overwritten to have Variable set.
* `sparseTensorSetToDeviceType` and similar functions previously did an equality test with TensorTypeId, to determine what an appropriate device type is. This equality is now replaced with a set inclusion test. This is valid, under the assumption that we don't ever have weird sets like "this tensor is simultaneously a sparse CPU tensor and a sparse CUDA tensor", which will be true in the short term plan of adding Variable to the dispatch ID.
* `impl::dispatchTypeId` was generally introduced for cases where we legitimately need to convert from `TensorTypeSet -> TensorTypeId` in a dispatch related manner. At the moment, the implementation is trivial, but they will soon be adjusted to handle TLS. I've tried to make these call sites as forwards compatible as possible:
  * `checked_tensor_unwrap` and co now use `dispatchTypeId`. When Variable is added to the type set, these will always be called in a context where the Variable type ID is disabled, so we will get the correct underlying tensor type ID.
  * Uses of `Backend` in dispatch are now replaced with `TensorTypeSet`. The general heuristic here for whether or not to accept a `TensorTypeId` or `TensorTypeSet` is that we want to make the generated code as simple as possible. It is easier to retrieve a `TensorTypeSet`, so that's a more appropriate API in these cases.
* In some cases, I could not conveniently switch an implementation to the new semantics, because it was blocked on some other refactor. In this case, I introduced `legacyExtractTypeId`, which gives what would be a BC-compatible `TensorTypeSet` to `TensorTypeId` implementation that will continue to report the same values it would have prior to this change. This is **different** from `dispatchTypeId`, because this function does NOT respect TLS; it always ignores Variable type IDs.
  * c10 dispatcher tests, which are oblivious to Variable dispatch, use this BC function (actually, they use `extractTypeId`, an overload for Tensor.
  * The implementation of `new_*` methods heavily relies on tensor type ID, I chose not to unwind this. PR to refactor this at https://github.com/pytorch/pytorch/pull/25475
  * Slicing also relies on tensor type ID, see `torch/csrc/autograd/python_variable_indexing.cpp` (though in some cases in this file, I was able to replace use of tensor type ID with TensorOptions)
* In some cases, there is an equality test on tensor type ID which would be better done by testing "tensor axes". In those cases, I replaced those equality tests with more equality tests.
  * Example: `torch/csrc/nn/type_checks.h`
  * There is a total punt in `torch/csrc/tensor/python_tensor.cpp` where "instance of" checking is done via dispatch ids. In general, the Variable-ness of a tensor doesn't participate in instanceof testing. It's not entirely clear what to do here.
  * Instead of storing `Backend` in `VariableInfo`, we now just store Layout.

c10 dispatcher test updates were done with:

```
:%s/\([^ ]\+\)\.type_id()/extractTypeId(\1)/g
:%s/\([^( ]\+\)->type_id()/extractTypeId(*\1)/g
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25308

Differential Revision: [D17092791](https://our.internmc.facebook.com/intern/diff/D17092791)

ezyang commented:
It turns out killing the `TensorTypeSet` constructor is a little more tricky than I thought. The problem is we have a number of shallow copy operations which copy all of the fields, including the tensor type set. If the only constructor is `TensorTypeId`, it's non-obvious WHICH field in the tensor type set should get copied over; you'd basically have to use `legacyExtractTensorId`. This seems a bit bad. Maybe what the code is telling me is that we should make a distinction between the "wrapper" section of the tensor type set, and the "kernel" section of the tensor type set.

gchanan commented:
> It turns out killing the TensorTypeSet constructor is a little more tricky than I thought. The problem is we have a number of shallow copy operations which copy all of the fields, including the tensor type set. If the only constructor is TensorTypeId, it's non-obvious WHICH field in the tensor type set should get copied over; you'd basically have to use legacyExtractTensorId. This seems a bit bad. Maybe what the code is telling me is that we should make a distinction between the "wrapper" section of the tensor type set, and the "kernel" section of the tensor type set.

I think it's also fine to have both constructors and just minimize the use of the TensorTypeSet constructors for the case you mentioned.
